### DISCUSSION TITLE
Thoughts on configuring the DB to platforms like Turso?

### DISCUSSION THREAD

**Discussion by gothammm** [pocketbase/pocketbase#2638] [Q&A]
This is a very rough idea, haven't thought through the details, but was hoping to get initial feedback on thoughts of allowing integration with platforms like https://turso.tech/

**Comment by ganigeorgiev** [comment]
I don't think this can be done easily. PocketBase by default can work only on a single server.

If you need replication you can check tools like [litestream](https://litestream.io/), [litefs](https://github.com/superfly/litefs), [marmot](https://github.com/maxpert/marmot), but again keep in mind that PocketBase wasn't designed with distributed usage in mind and some things may not work as expected (eg. the realtime events and the app hooks in general).

Support for external SQL databases at least for now is out of the scope for v1 (_if you haven't already, you can explore the active tasks in the [roadmap](https://github.com/orgs/pocketbase/projects/2/views/1)_).

**Comment by rnbokade** [comment]
I think this can be done, by replacing the sqlite with libsql. Libsql is mostly drop in replacement for sqlite. where you specify the target url instead of target file.

**Comment by rnbokade** [comment]
https://github.com/libsql/libsql-client-go

**Comment by richardvanbergen** [comment]
Personally I'd like to see some support for cloud infra. I think it's good for everything to be local by default but I'm not an infra guy and I don't really want to have to think about backups and stuff. I don't trust myself with this either.

It would be a great feature to be able to connect to a remote provider for data. The whole point of pocketbase is I want to focus on the application and I think this aligns with that goal.

**Reply by rnbokade** [reply]
Now with recrnt libsql release this has become more convinent... there is a new mode where local db file behaves as it is and turso acts like a replication target. We just need to swap drivers now

**Reply by fredguth** [reply]
I would love to see a repo trying this idea.

**Comment by danfascia** [comment]
I would love to see support for Turso or other libSQL based solutions.

Turso would open up great possibilities for pocketbase and I don't see it being a problem since the writes in Turso always happen to a single master which is then replicated geographically by the Turso service.

It would also just be really cool to run the SQLite instance in a separate location to the server pocketbase runs on to allow separate scaling of storage vs compute.

**Reply by hugo-abdou** [reply]
i wish if this comes true

**Comment by MengLinMaker** [comment]
I am currently exploring how to combine PocketBase with Turso.
This would require an embedded Turso (read only) replica (which replicates on demand from Turso primary database).

Unfortunately any writes from PocketBase will not be updated in the Turso primary database.
Unless the SQL statements are streamed to Turso SDK.

With this constraint, one solution is to extend PockeBase to use Turso SDK on all database writes - not sure if the db hooks cover all writes, updates...
Another solution is to stream PocketBase SQL queries in dev mode then update Turso from the stream.

Both solutions so far seem kinda janky.

**Comment by MengLinMaker** [comment]
So here's a hacky package that streams SQL from PocketBase to Turso:
https://github.com/MengLinMaker/TursoBase
Still more testing to do.

Maintainer insists that PocketBase is not scalable which is fair enough. So probably some features to not work.

I'm only interested in authentication with PocketBase.

**Comment by ganigeorgiev** [comment]
I'm closing this discussion as it starting to become a little spammy.

As mentioned in other similar posts, with the [planned refactoring](https://github.com/pocketbase/pocketbase/discussions/4355) users will be able to define their own db connect function to load custom SQLite builds and drivers (including Turso/libsql).

But the main point from this discussion remains - PocketBase is not designed to scale horizontally and if your use case requires it you'll be better with something else.

Personally for me the only valid use case for using PocketBase with Turso is not "scaling" but the autosync/replication for disaster recovery (which you could already achieve with Litestream or other SQLite streaming replication tools).

**Reply by RickCastle2018** [reply]
Scaling horizontally is one question, but what if we'd want, at some point to add a data analyzing micro service to the system? For me, it's okay that PocketBase itself does not scale, but the database could be scaled enlarging PBs' use case - any successful app grows at some point. Among competitors PB is the best solution, the only Go solution. The problem is 'what's when we grow'?

**Reply by ganigeorgiev** [reply]
I don't understand your example with "data analyzing micro service" and what you are trying to say. As mentioned on the website PocketBase is suitable for small and medium sized projects that can operate on a single server. If your application grows beyond that then you rewrite it. At that point your business requirements would have changed at least several times since you've started the project and you'll be better (and will have the resources) to build a new version of your app from scratch accommodating your specific use case.

If you are not satisfied with PocketBase you can simply use something else. I don't know from where this entitlement come from and why the users of a personal open source project behave like paying customers and expects anything more than what is already published, but I find this type of discussions pointless and will stop engage further. As mentioned in the above comment with the refactoring and the upcoming v0.23.0 users will be able to define their own db connect function to load custom SQLite drivers (see for example https://github.com/pocketbase/pocketbase/discussions/5611).
